### PR TITLE
Fix UI and backend paths with subpath

### LINKED ISSUE #1773: use  'WOODPECKER_ROOT_URL'   can't  login
### Component

server

### Describe the bug

when i use `WOODPECKER_ROOT_URL`,  woodpecker  can't  login 

```
 apply style from 'http://xxx/ci/assets/index-6d181f24.css' because its MIME type ('text/html') is not a supported stylesheet MIME type, and strict MIME checking is enabled.
ci:1  Refused to execute script from 'http://xxx/ci/web-config.js' because its MIME type ('text/html') is not executable, and strict MIME type checking is enabled.
index-b6d7ab2a.js:1  Failed to load module script: Expected a JavaScript module script but the server responded with a MIME type of "text/html". Strict MIME type checking is enforced for module scripts per HTML spec.
ci:1  Refused to apply style from 'http://xxx/ci/assets/index-6d181f24.css' because its MIME type ('text/html') is not a supported stylesheet MIME type, and strict MIME checking is enabled.
```

### System Info

```shell
kubernets 1.25

woodpecker v0.15.9 
woodpecker image tag：next
```

### Additional context

_No response_

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't [already an issue](https://github.com/woodpecker-ci/woodpecker/issues) that reports the same bug to avoid creating a duplicate.
- [X] Checked that the bug isn't fixed in the `next` version already [https://woodpecker-ci.org/faq#which-version-of-woodpecker-should-i-use]
- [X] Check that this is a concrete bug. For Q&A join our [Discord Chat Server](https://discord.gg/fcMQqSMXJy) or the [Matrix room](https://matrix.to/#/#woodpecker:matrix.org).

### LINKED ISSUE #1798: use 'WOODPECKER_ROOT_URL' not operating properly
### Component

server

### Describe the bug

i see https://github.com/woodpecker-ci/woodpecker/pull/1783  pr。
When I used the tag of [next-f91ee5d23a](https://hub.docker.com/layers/woodpeckerci/woodpecker-agent/next-f91ee5d23a/images/sha256-bf85a8ad052d42364d96214e8cbf45fe789ba185600d9018124c2ac3c818e268?context=explore)   to verify, I found that the function of `WOODPECKER_ROOT_URL` had not been developed. I used woodpecker + gitea and only corrected the access path of the front-end static file display at present. The other URIs are not deeply modified, https://github.com/devture/com.devture.ansible.role.woodpecker_ci_server/pull/2#issuecomment-1530932964 will this issue on the agenda

### System Info

```shell
kubernets 1.25

woodpecker next-f91ee5d23a 

gitea: latest
```

### Additional context

_No response_

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't [already an issue](https://github.com/woodpecker-ci/woodpecker/issues) that reports the same bug to avoid creating a duplicate.
- [X] Checked that the bug isn't fixed in the `next` version already [https://woodpecker-ci.org/faq#which-version-of-woodpecker-should-i-use]
- [X] Check that this is a concrete bug. For Q&A join our [Discord Chat Server](https://discord.gg/fcMQqSMXJy) or the [Matrix room](https://matrix.to/#/#woodpecker:matrix.org).

### PR DESCRIPTION
**Pull request by qwerty287**
I'm not sure if this is an ideal fix for this, but it seems to work for me. If you have another idea just let me know.

Closes #1798 
Closes #1773

### CODE REVIEW

File: web/src/router.ts [hunk 1]
```diff
@@ -2,34 +2,39 @@ import { Component } from 'vue';
 import { createRouter, createWebHistory, RouteRecordRaw } from 'vue-router';
 
 import useAuthentication from '~/compositions/useAuthentication';
+import useConfig from '~/compositions/useConfig';
 import useUserConfig from '~/compositions/useUserConfig';
 
+let { rootURL } = useConfig();
+if (rootURL !== '' && rootURL !== null) {
```
Comment (liyimeng) [inline web/src/router.ts:9]: Using  '??' to chech  rootURL can be more robust?

**Comment by Seryta** [comment]
Some links in js still have no rooturl, and this pr still doesn't solve the problem for me.
image

**Comment by Seryta** [comment]
> Can you give the exact issues? Were are links that do not work for you?

https://github.com/woodpecker-ci/woodpecker/issues/1798#issuecomment-1573191668

**Comment by Seryta** [comment]
> @Seryta I tried fixing them in [0ce8f1f](https://github.com/woodpecker-ci/woodpecker/commit/0ce8f1fde805cd8f6a516aa62eb00b3ec923cad8). It's possible that I missed something, can you test again?

this is the latest situation: https://github.com/woodpecker-ci/woodpecker/issues/1798#issuecomment-1573233074

**Comment by Seryta** [comment]
> Are there any errors in the browser/JS console?

Oh, yes, the error is as follows:
image

image

**Comment by Seryta** [comment]
There may also be a problem with another file, web/src/lib/api/index.ts

https://github.com/woodpecker-ci/woodpecker/blob/71d7a467f904eb83432227450693f1b66b6075dd/web/src/lib/api/index.ts#L290

https://github.com/woodpecker-ci/woodpecker/blob/71d7a467f904eb83432227450693f1b66b6075dd/web/src/lib/api/index.ts#L303

**Comment by Seryta** [comment]
> Can you try again with [75ef337](https://github.com/woodpecker-ci/woodpecker/pull/1799/commits/75ef3372d8bca23f88c7b8e815f7d66b7c9063c9)?

Now the content can be displayed, but as I said above, there is a problem with `/stream/events` access, and `/assets/logo-c3ebba12.svg` or `/assets/logo-c3ebba12.svg`, etc., are still Links in `index-da1e4087.js` do not have root_url.

In addition, what is more serious is that when performing oauth2 authorization authentication, the link to jump is `/authorize` instead of `/ci/authorize`, and the link to jump after authorization is not `/ci` but `/`

**Comment by Seryta** [comment]
> 

I'm using gitea, and the authorization jump link is indeed `example.com/ci/authorize`, but it doesn't work, I'm not sure where the problem is

**Comment by Seryta** [comment]
> @Seryta Can you please check again with the latest commit. I tried to fix both streaming and redirecting, not sure if I found everything though.

@qwerty287 It's great, but there are still some problems with some pages. First, the login access is still `/authorize` instead of `/ci/authorize`, and secondly, `/assets` still has no root_path after logging in.

image

**Comment by Seryta** [comment]
After manually modifying the url to access `/ci/authorize`, there is no problem with the link that jumps back after the authorization is successful.

File: server/router/router.go [hunk 2]
```diff
@@ -16,10 +16,12 @@ package router
 
 import (
 	"net/http"
+	"net/url"
```
Comment (Seryta) [inline server/router/router.go:19]: please delete it, "net/url" imported and not used

File: server/web/web.go [hunk 3]
```diff
@@ -53,15 +53,14 @@ func New() (*gin.Engine, error) {
 
 	e.Use(setupCache)
 
-	rootURL, _ := url.Parse(server.Config.Server.RootURL)
```
Comment (Seryta) [inline server/web/web.go:56]: please remove "net/url"

**Review by Seryta — CHANGES_REQUESTED** [review]

**Comment by liyimeng** [comment]
@qwerty287 shall we squash the commits into a single one? it is kind of messy

**Comment by Seryta** [comment]
> @Seryta I looked again for usages without root path, can you try again with the latest commit? I'm really sorry to bother you for testing this.

Sorry, I just saw it today, I tested it, and the problem mentioned above still exists.
1. Automatically access `/authorize` instead of `/ci/authorize` when logging in for authorization
2. After manually accessing `/ci/authorize`, there is a problem with accessing the following links:
+ https://example.com/ci/assets/logo.svg?url, 404
+ https://example.com/assets/Repos-2587f056.js, 404
+ https://example.com/assets/ListItem.vue_vue_type_script_setup_true_lang-dc8e4876.js, 404
+ https://example.com/assets/Scaffold.vue_vue_type_script_setup_true_lang-4e0e2911.js, 404
+ https://example.com/assets/FluidContainer.vue_vue_type_script_setup_true_lang-df900462.js, 404
+ https://example.com/assets/useRepoSearch-2903cf6a.js, 404

**Comment by 6543** [comment]
@liyimeng pulls are squash-merged anyway ...

**Comment by liyimeng** [comment]
> @liyimeng pulls are squash-merged anyway ...

do you mean it will squashed at merge automatically, cool !

**Comment by lafriks** [comment]
For fronted part that needs to support subpaths I compile in vite with base path `<BASE_PATH>` and than on app startup replace it with configured value in all static files.

**Review by 6543 — CHANGES_REQUESTED** [review]
logo is broken if WOODPECKER_ROOT_PATH=""

**Comment by 6543** [comment]
html:
```
<img class="w-48 h-48" src="../assets/logo.svg?url">
```

**Comment by 6543** [comment]
login screen is still a white page :/

**Comment by 6543** [comment]
just as an idea ... instead of this "catch-all" aproach ... what if we just introduce a "basePath" that is `""` by default ... and can be changed so it will reflect this automatically ... and on the UI site we can use relative paths instead of absulte ... that should solve it ?

**Comment by lafriks** [comment]
> Unfortunately, this is not possible because of Vite's compiling system which assumes that you use a prefix that is set at compile time. I didn't see any option to change this.

Usually I use something like `{{BASEPATH}}` in vite and then either with shell script in nginx docker image or with go code if embedded in binary replace it with actual subpath or `/` on start in index.html and *.js/*.css files.

**Comment by xoxys** [comment]
Unpopular opinion: Remove any support for subpath deployments, as this always leads to problems. What are the benefits? Nowadays you can get certificates for free, and subdomains are not charged for in most cases either.

**Comment by gmemstr** [comment]
To counter that, I'm very eager to see this working because I personally use Tailscale Funnel to expose some services to the internet. Tailscale Funnel only exposes a single subdomain for a machine, and so everything has to be routed with subpaths. Other services I use handle this fine (e.g Gitea/Forgejo, n8n etc) and _not_ supporting it is how I ended up opting for Gitea Actions instead. I can guarantee you I'm not the only one in this sort of position (and heck, maybe it's just for aesthetic purposes, but regardless, I believe that is it something that should & can work).

**Review by 6543 — APPROVED** [review]

**Pull request merged**